### ISSUE TITLE
add targets in function build_targets

### ISSUE THREAD

**Issue by Laughing-q** [ultralytics/yolov5#461]
## ❔Question
@glenn-jocher  hey glenn, I have seen some code recently, and I found that you added some overlap target boxes for training by adding offset(function build_targets in utils/utils.py). What's the benefit of doing so? I haven't figured out yet.

**Comment by Laughing-q** [comment]
code details:


    style = 'rect4'
    for i in range(det.nl):
        anchors = det.anchors[i]
        gain[2:] = torch.tensor(p[i].shape)[[3, 2, 3, 2]]  # xyxy gain
        # Match targets to anchors
        a, t, offsets = [], targets * gain, 0
        if nt:
            r = t[None, :, 4:6] / anchors[:, None]  # wh ratio
            j = torch.max(r, 1. / r).max(2)[0] < model.hyp['anchor_t']  # compare
            # j = wh_iou(anchors, t[:, 4:6]) > model.hyp['iou_t']  # iou(3,n) = wh_iou(anchors(3,2), gwh(n,2))
            a, t = at[j], t.repeat(na, 1, 1)[j]  # filter

            # overlaps
            gxy = t[:, 2:4]  # grid xy
            z = torch.zeros_like(gxy)
            if style == 'rect2':
                j, k = ((gxy % 1. < g) & (gxy > 1.)).T
                a, t = torch.cat((a, a[j], a[k]), 0), torch.cat((t, t[j], t[k]), 0)
                offsets = torch.cat((z, z[j] + off[0], z[k] + off[1]), 0) * g
            elif style == 'rect4':
                j, k = ((gxy % 1. < g) & (gxy > 1.)).T
                l, m = ((gxy % 1. > (1 - g)) & (gxy < (gain[[2, 3]] - 1.))).T
                a, t = torch.cat((a, a[j], a[k], a[l], a[m]), 0), torch.cat((t, t[j], t[k], t[l], t[m]), 0)
                offsets = torch.cat((z, z[j] + off[0], z[k] + off[1], z[l] + off[2], z[m] + off[3]), 0) * g

**Comment by seekFire** [comment]
@Laughing-q 
Thanks for your careful research! I'm also concerned about this issue.
BTW, I'm also curious about what's the meaning of **targets[:, 0]**. as you know in **build_targets**, [this line](https://github.com/ultralytics/yolov5/blob/5e970d45c44fff11d1eb29bfc21bed9553abf986/utils/utils.py#L548) shows that b = t[:, 0], the annotation nearby said it is 'image', but in dataset.py, [the line](https://github.com/ultralytics/yolov5/blob/5e970d45c44fff11d1eb29bfc21bed9553abf986/utils/datasets.py#L551) definiting the **targets** shows that the targets[:, 0] is always equal to zero, so I can't understand what on earth is the meaning of targets[:, 0]

**Comment by Laughing-q** [comment]
> BTW, I'm also curious about what's the meaning of **targets[:, 0]**. as you know in **build_targets**, [this line](https://github.com/ultralytics/yolov5/blob/5e970d45c44fff11d1eb29bfc21bed9553abf986/utils/utils.py#L548) shows that b = t[:, 0], the annotation nearby said it is 'image', but in dataset.py, [the line](https://github.com/ultralytics/yolov5/blob/5e970d45c44fff11d1eb29bfc21bed9553abf986/utils/datasets.py#L551) definiting the **targets** shows that the targets[:, 0] is always equal to zero, so I can't understand what on earth is the meaning of targets[:, 0]

@seekFire As you know, a image has not only one target and a batch has multiple pictures,  targets[:, 0] is a index that helps us to distinguish which target belongs to which image in a batch.
 targets in a batch like this:
[[0, 6, 0.5, 0.5, 0.26, 0.35],
     [0, 6, 0.5, 0.5, 0.26, 0.35],
     [1, 6, 0.5, 0.5, 0.26, 0.35],
     [2, 6, 0.5, 0.5, 0.26, 0.35],]
The first two lines of labels belong to the first(index 0) picture, and the third line belongs to the second.
And setting targets[:, 0] is in [this function](https://github.com/ultralytics/yolov5/blob/5e970d45c44fff11d1eb29bfc21bed9553abf986/utils/datasets.py#L562) 

But I haven't figured the issue(add extra targets) out yet.

**Comment by seekFire** [comment]
@Laughing-q Wow, I got it... the question rised by my unfamiliarity with pytorch, thank you very much! BTW, I will also wait together with you for the answer of this issue

**Comment by zugofn** [comment]
> ## grey_questionQuestion
> 
> @glenn-jocher hey glenn, I have seen some code recently, and I found that you added some overlap target boxes for training by adding offset(function build_targets in utils/utils.py). What's the benefit of doing so? I haven't figured out yet.

Hello,maybe you can see this blog for your question:https://blog.csdn.net/cdknight_happy/article/details/109817548#t11
